### ISSUE TITLE
Include and use Argon2 password hashing on Retroshare

### ISSUE THREAD

**Issue by ASmith-** [RetroShare/RetroShare#1367]
Include and use Argon2 password hashing on Retroshare

[Tascloud](https://tascloud.me) will begin shifting its password hashing and other hashing tasks over to include Argon2 in the months ahead. Argon2 is fast, optimized, secure against multiple attacks and still unique. 

4yrs ago I began a spirited discussion in the Retroshare signed Developers chat lobby and developers forums urging updating the password hash algorithm used by Retroshare from the old tired SHA2 algorithm over to Argon. And posted at length the advantages of doing so over SHA2. The lead Retroshare developers should now finally consider updating the hash algorithm from Sha2 to Argon2 on Retroshare to provide further security, further hashing speeds and other positives that come with that hashing upgrade.

Argon2 source code package is distributed under the Creative Commons CC) 1.0 License

References
https://github.com/khovratovich/Argon2
[Argon2-master.zip](https://github.com/RetroShare/RetroShare/files/2458253/Argon2-master.zip)

https://www.cryptolux.org/images/0/0d/Argon2.pdf
[Argon2.pdf](https://github.com/RetroShare/RetroShare/files/2458254/Argon2.pdf)

https://www.cryptolux.org/index.php/Argon2#Winner_of_PHC
Argon2 is a new hash function, which summarizes the state of the art in the design of memory-hard functions. It is a streamlined and simple design. It aims at the highest memory filling rate and effective use of multiple computing units, while still providing defense against tradeoff attacks. Argon2 is optimized for the x86 architecture and exploits the cache and memory organization of the recent Intel and AMD processors.

**Comment by felisucoibi** [comment]
Even if this is done in the future, should be retrocompatible with sha1, important to take care of this.

**Comment by ASmith-** [comment]
Hi felisucoibi, unfortunately the gain in file sharing security would be lessened by being legacy backwards compatible with sha1 hashing which state sponsored attackers, and crackers have large databases of files with their sha1 hash sums to target instead of merely looking for the file names.

If a Retroshare developer had previously developed a nifty external .db database file compatible with 3rd party external databases for Retroshare Users to pull up the shared files of themselves and their friends by filename and sha1 sums, this lack of backward compatibility would be more of a issue IMO. However, lacking that effort by the Retroshare developers, a clean break and using the generation 3+ hashing ultra-fast, secure and optimized Argon2 would make a very good addition to Retroshares source code incorporating the C++ source coding which I've already provided.

Moreover the gain in password hashing security and other areas would be amount to well over a full generation in optimization and attack proofing from SHA2 Sha-256 to SHA3+ on Argon2.

**Comment by felisucoibi** [comment]
The legacy is important to not loose all the content inside rs newtwork, i'm sure could be  secure way to keep old the legacy content and use the new hashes for new content and in the future remove sha1, anyway by now is not a priority in rs i think.

**Comment by ASmith-** [comment]
Yes, I agree felisucoibi , there's many thousands of .rscollections which depend on sha1 hashsum file values on the global Retroshare network. A hybrid use until RS 0.7 is released would provide enough time and momentum for that change over. Argon2 provides far more than hashing optimizations, also password and encryption security as well.
